sdk/nodejs/errors.ts

```diff
@@ -81,3 +81,16 @@ export function isGrpcError(err: Error): boolean {
     const code = (<any>err).code;
     return code === grpc.status.UNAVAILABLE || code === grpc.status.CANCELLED;
 }
+
+export class DetailedError extends Error {
+
+    public readonly __pulumiDetailedError: boolean = true;
+
+    constructor(message: string, public details?: Array<[string, string]>) {
+	super(message);
+    }
+
+    public static isInstance(obj: any): obj is DetailedError {
+	return utils.isInstance<DetailedError>(obj, "__pulumiDetailedError");
+    }
+}
```

Comment:
I think there are a few options here:

* Have a single error class that covers all kinds of details like the suggestion here
* Have an error class per status/case we care about
* Have a single error class that can handle several cases (a bit like tRPC's `TrpcError` -- https://trpc.io/docs/server/error-handling)

I'm leaning towards the third, then second, then first in order of preference personally. I feel like the user wants to write something like this:

```typescript
// Class per status
throw new InvalidArgumentError({
  message: "...",
  fields: [
    { name: "foo", description: "foo was too small" },
    { name: "bar", description: "bar was too big" },
  ]
})

// Single type with discriminator
throw new PulumiException({
  type: "INVALID_ARGUMENT",
  message: "...",
  fields: [ ... /* as before */ ...],
})
```

We could make fields optional or mandatory as we'd like; it might make sense to have a `cause` too and take the message/stack off that if it works, but that could easily be added later I reckon.

I would say whatever we do I'd prefer objects to tuples, and probably for compound arguments in general -- they will I think serve us better when it comes to backwards compatibility (and IMO are a bit more readable, but that's more subjective I appreciate). Just my 2c anyway!

Reply:
Agree, although I think I prefer option 2 over 3 in terms of feeling nicer to use. Again repeating what I said in ideation the core thing is whatever the engine understands needs to be mirrored into the SDK to send. If the engine can take N status codes and treat the different the SDK needs to be able to send all of them.

Reply:
> So the premise here is that gRPC status codes don't make a lot of sense in the context of Pulumi

Yeh... I wonder if we're being mad trying to fit googles rpc error structure into our engine. Like the basic concept here (return a richer error struct via grpc) is sound, but maybe rather than importing googles grpc error details we should just add our own? Then we can make it shaped in a way that totally fits the engine.